src/ripple/app/reporting/ETLHelpers.h

```diff
@@ -0,0 +1,194 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_ETLHELPERS_H_INCLUDED
+#define RIPPLE_CORE_ETLHELPERS_H_INCLUDED
+#include <ripple/app/main/Application.h>
+#include <ripple/ledger/ReadView.h>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <sstream>
+
+namespace ripple {
+
+/// This datastructure is used to keep track of the sequence of the most recent
+/// ledger validated by the network. There are two methods that will wait until
+/// certain conditions are met. This datastructure is able to be "stopped". When
+/// the datastructure is stopped, any threads currently waiting are unblocked.
+/// Any later calls to methods of this datastructure will not wait. Once the
+/// datastructure is stopped, the datastructure remains stopped for the rest of
+/// it's lifetime.
+class NetworkValidatedLedgers
+{
+    // max sequence validated by network
+    std::optional<uint32_t> max_;
+
+    std::mutex mtx_;
+
+    std::condition_variable cv_;
+
+    bool stopping_ = false;
+
+public:
+    /// Notify the datastructure that idx has been validated by the network
+    /// @param idx sequence validated by network
+    void
+    push(uint32_t idx)
+    {
+        std::unique_lock<std::mutex> lck(mtx_);
```

Comment:
Drop `<std::mutex>` (throughout file).
In this place, `std::lock_guard` is better because it can't let go of the lock until destruction.

Reply:
Actually, it looks like we can't pass `std::lock_guard` to `std::condition_variable::wait()`: 
https://en.cppreference.com/w/cpp/thread/condition_variable/wait
https://stackoverflow.com/questions/40577622/is-possible-to-use-stdcondition-variable-with-stdlock-guard

Reply:
That is correct.  But within `push` and `close`, you don't call `wait`.  Certainly in `getMostRecent`, retain `unique_lock` and just drop the `<mutex>` there.

Reply:
oh yes, you're right. Don't know how I missed that.

Reply:
changed to `std::lock_guard` everywhere that we don't `wait`. Dropped `std::mutex` from the places where we do `wait`